---
title: "chipPeakAnno_demo"
output: html_document
---

```{r, message=FALSE}
library(Seurat)
library(Signac)
library(ChIPpeakAnno)
library(dplyr)
library(BSgenome.Hsapiens.UCSC.hg38)
library(TxDb.Hsapiens.UCSC.hg38.knownGene)
library(EnsDb.Hsapiens.v86)
library(biomaRt)
library(plyranges)
library(motifStack)
```

# Compare replicates

```{r, warning=FALSE}
# load data in
fli1_macs2_r1 <- toGRanges("../../../ewrenn/smad_cutnrun/results/04_reporting/igv/TGFB_FLI1_R1.macs2.peaks.cut.bed")
fli1_macs2_r2 <- toGRanges("../../../ewrenn/smad_cutnrun/results/04_reporting/igv/TGFB_FLI1_R2.macs2.peaks.cut.bed")

# overlap peaks
overlaps <- findOverlapsOfPeaks(fli1_macs2_r1, fli1_macs2_r2) # overlaps two GRanges objects
```


```{r, fig.width=5, fig.height=5, message=FALSE}
# how do replicates compare?
venn <- makeVennDiagram(overlaps, 
                        # totalTest = 100000,
                        fill = c("#009E73", "#F0E442"),
                        col = c("#D55E00", "#0072B2"),
                        cat.col = c("#D55E00", "#0072B2"))

```


```{r, warning=FALSE}
# prepare annotation (i.e. hg38 genes)
ensembl.hs86.gene <- toGRanges(EnsDb.Hsapiens.v86, feature = "gene")

# how close are peaks to transcription start sites?
binOverFeature(overlaps$mergedPeaks, 
               nbins = 20,
               annotationData = ensembl.hs86.gene,
               xlab = "peak distance from TSSs (bp)", 
               ylab = "peak count", 
               main = "Distribution of aggregated peak numbers around TSS")
```


```{r, fig.width=5, fig.height=5, message=FALSE}
# how do the peaks overlap with each other? (part 4)
# uses overlappingPeaks object
unique(overlaps$overlappingPeaks[["fli1_macs2_r1///fli1_macs2_r2"]][["overlapFeature"]])
pie1(table(overlaps$overlappingPeaks[["fli1_macs2_r1///fli1_macs2_r2"]][["overlapFeature"]]))
# pie1(table(overlaps$mergedPeaks))

```


```{r, fig.width=6, fig.height=6, warning=FALSE}
# what genomic features correlate with the peaks? (part 6)
# uses GRanges objects
macs_peaks <- GRangesList(rep1 = fli1_macs2_r1, rep2 = fli1_macs2_r2)
genomicElementDistribution(macs_peaks, TxDb = TxDb.Hsapiens.UCSC.hg38.knownGene)
```


```{r, message=FALSE, warning=FALSE}
# annotate peaks using hg38 genome
fli1_total_anno <- annotatePeakInBatch(overlaps$mergedPeaks, 
                                       AnnotationData = transcripts(EnsDb.Hsapiens.v86))

fli1_promoter_anno <- annotatePeakInBatch(overlaps$mergedPeaks, # worth checking ?annotatePeakInBatch
                                       AnnotationData = ensembl.hs86.gene,
                                       output = "nearestBiDirectionalPromoters",
                                       bindingRegion = c(-2000, 2000)) # output parameter is important

peaks_near_bdp <- peaksNearBDP(overlaps$mergedPeaks,
                               AnnotationData = ensembl.hs86.gene,
                               MaxDistance = 2000)

# clean up resulting object, ensure all rows are unique
fli1_promoter_anno <- unname(fli1_promoter_anno) 
fli1_promoter_anno$peakNames <- NULL 
```


```{r}
pie1(table(fli1_total_anno$insideFeature)) + title("all peaks")
pie1(table(fli1_promoter_anno$insideFeature)) + title("promoter - associated peaks")
```


```{r}
# if you want to get the sequence of the peaks, used for motif analysis
sequence_around_peak <- getAllPeakSequence(overlaps$mergedPeaks, 
                                           upstream = 20,
                                           downstream = 20, 
                                           genome = BSgenome.Hsapiens.UCSC.hg38)

freqs <- oligoFrequency(BSgenome.Hsapiens.UCSC.hg38$chr1) # frequency of oligos in hg38 chr 1
motif_summary <- oligoSummary(sequence_around_peak, 
                              freqs = freqs, 
                              quickMotif = TRUE
                              )
zscore <- sort(motif_summary$zscore)
```


```{r}

pfm <- new("pfm", mat = motif_summary$motifs[[1]],
           name = "sample motif 1")
motifStack(pfm)

```

# Compare multiple binding profiles

```{r}
CHLA10 <- readRDS("../working_data/CHLA10_subset.rds")
```


```{r, warning=FALSE}
# get DA ranges and genes
DefaultAssay(CHLA10) <- "ATAC"
Idents(CHLA10) <- "condition"
da_ranges <- FindMarkers(CHLA10, ident.1 = "TGFB1", ident.2 = "vehicle")
da_ranges <- subset(da_ranges, da_ranges$p_val_adj <= 0.05)
da_genes <- ClosestFeature(CHLA10, regions = rownames(da_ranges))
```


```{r}
# create GRanges object from DA peak ranges
da_coords <- rownames(da_ranges)
coords_parts <- strsplit(da_coords, split = "-")
coords_df <- data.frame(
  seqnames = sapply(coords_parts, `[`, 1),
  start = as.numeric(sapply(coords_parts, `[`, 2)),
  end = as.numeric(sapply(coords_parts, `[`, 3))
)
da_pks <- as_granges(coords_df)
```


```{r}
# make a venn diagram to check peak overlaps
cross_pks <- findOverlapsOfPeaks(overlaps$mergedPeaks, da_pks) # overlap DA peaks with FLI1 C&R peaks

mean_peak_width <- mean(width(unlist(GRangesList(cross_pks[["all.peaks"]]))))
total_binding_sites <- length(BSgenome.Hsapiens.UCSC.hg38[["chr2"]]) * 0.03 / mean_peak_width
venn1 <- makeVennDiagram(cross_pks, 
                         # totalTest = total_binding_sites, 
                         fill = c("#CC79A7", "#56B4E9"),
                         col = c("#D55E00", "#0072B2"),
                         cat.col = c("#D55E00", "#0072B2"))

# venn1[["p.value"]] # check significance of overlap
# venn1[["vennCounts"]] # check counts of overlap (goes into venn diagram already)
```


```{r}
# center peaks 
peak_center <- reCenterPeaks(cross_pks$mergedPeaks, width = 4000) 

# set path to bigwig files
bw_path <- "../../../ewrenn/smad_cutnrun/results/04_reporting/igv"
bigWigs_all <- dir(bw_path, "bigWig")
bigWigs <- bigWigs_all[c(2,4,6)]
```



```{r}
# combine centered peaks with C&R bigwig signal
coverage_list <- sapply(file.path(bw_path, bigWigs), 
                        import, # rtracklayer::import
                        format = "BigWig",
                        which = peak_center,
                        as = "RleList")
names(coverage_list) <- gsub(".bigWig", "", bigWigs) # fix names

sig <- featureAlignedSignal(coverage_list, peak_center) # generate overlapping signal

keep <- rowSums(sig[[1]]) > 0 & rowSums(sig[[2]]) > 0 & rowSums(sig[[3]]) > 0 # keep peaks that have signal for all three TFs
sig <- sapply(sig, function(x) x[keep, ], simplify = FALSE) # apply to signal
peak_center <- peak_center[keep] # apply to peaks


```


```{r}
# merged figs w/ bigwigs; shows TGFB x FLI1 overlap peak signal on CNR TF binding sites
featureAlignedHeatmap(sig, peak_center,
                      upper.extreme=c(3, 1.75, 2.5))
featureAlignedDistribution(sig, peak_center, 
                           type="l")

```



