---
title: "peaks_atac_cnr"
output: html_document
---

```{r, message=FALSE}
library(Seurat)
library(Signac)
library(ChIPpeakAnno)
library(plyranges)
library(EnsDb.Hsapiens.v86)
library(biomaRt)
library(UpSetR)
library(TxDb.Hsapiens.UCSC.hg38.knownGene)
```


```{r}
# vtp_tgfb <- readRDS("../source_data/vtp_tgfb_multi.rds")
# Idents(vtp_tgfb) <- "cell_line"
# CHLA10 <- subset(vtp_tgfb, vtp_tgfb$cell_line == "CHLA10")
# saveRDS(CHLA10, "../working_data/CHLA10_subset.rds")
CHLA10 <- readRDS("../working_data/CHLA10_subset.rds")
fli1_peaks <- read_bed_graph("../../../ewrenn/smad_cutnrun/results/03_peak_calling/04_called_peaks/seacr/TGFB_FLI1_R1.seacr.peaks.stringent.bed")
```


```{r}
DefaultAssay(CHLA10) <- "ATAC"
Idents(CHLA10) <- "condition"
da_ranges <- FindMarkers(CHLA10, ident.1 = "TGFB1", ident.2 = "vehicle")
da_ranges <- subset(da_ranges, da_ranges$p_val_adj <= 0.05)
```


```{r, message=FALSE, warning=FALSE}
# annotate peaks on hg38 genome -- will be warning message since EnsDb is NCBI format, fli1 peaks are UCSC format
fli1_peak_annot <- annotatePeakInBatch(fli1_peaks, AnnotationData = transcripts(EnsDb.Hsapiens.v86))

# add gene symbols to annotation
mart <- useMart(biomart = "ENSEMBL_MART_ENSEMBL", dataset = "hsapiens_gene_ensembl")
fli1_peak_ensembl <- addGeneIDs(fli1_peak_annot, mart = mart, feature_id_type = "ensembl_transcript_id", IDs2Add = "hgnc_symbol")

# ensure all rows of object are unique
fli1_pks <- unique(fli1_peak_ensembl)
start(fli1_pks) <- as.numeric(start(fli1_pks))
end(fli1_pks) <- as.numeric(end(fli1_pks))

fli1_pks$start_position <- as.numeric(fli1_pks$start_position)

```


```{r}
# create gRanges object from DA peak ranges
da_coords <- rownames(da_ranges)
coords_parts <- strsplit(da_coords, split = "-")
coords_df <- data.frame(
  seqnames = sapply(coords_parts, `[`, 1),
  start = as.numeric(sapply(coords_parts, `[`, 2)),
  end = as.numeric(sapply(coords_parts, `[`, 3))
)
da_pks <- as_granges(coords_df)

```


```{r}
peak_overlaps <- findOverlapsOfPeaks(fli1_pks, da_pks)
names(peak_overlaps)
```


```{r, fig.width=5, fig.height=5}
# what proportions of the peaks overlap?
makeVennDiagram(peak_overlaps, totalTest = 100000, fill = c("#009E73", "#F0E442"),
                        col = c("#D55E00", "#0072B2"),
                        cat.col = c("#D55E00", "#0072B2"))
```


```{r, fig.width=5, fig.height=5}
# how do the peaks overlap with each other?
unique(peak_overlaps$overlappingPeaks[["fli1_pks///da_pks"]][["overlapFeature"]])
pie1(table(peak_overlaps$overlappingPeaks[["fli1_pks///da_pks"]][["overlapFeature"]]))
```


```{r, fig.width=6, fig.height=6}
# figure out how to interpret this

res <- genomicElementUpSetR(peaks = peak_overlaps[["peaklist"]][["fli1_pks///da_pks"]], 
                            TxDb.Hsapiens.UCSC.hg38.knownGene)

upset(res[["plotData"]], 
      nsets = length(colnames(res$plotData)), 
      nintersects = NA)

```


```{r}
# function to create genomic range for a gene of interest -- helpful for peak plots
goi_to_roi <- function(gene_name) {
  gene_granges <- genes(EnsDb.Hsapiens.v86, filter = GeneNameFilter(gene_name))
  if (length(gene_granges) == 0) {
    stop("Error: Gene '", gene_name, "' not found in EnsDb.Hsapiens.v86 annotations.")
  }
  gene_range_granges <- gene_granges[1]
  chr <- as.character(seqnames(gene_range_granges))
  start <- as.numeric(start(gene_range_granges))
  end <- as.numeric(end(gene_range_granges))
  gene_range_string <- paste0("chr", chr, "-", start, "-", end)
  return(gene_range_string)
}
```


```{r}
idents.plot <- c("vehicle", "TGFB1", "TGFB2")

covplot <- CoveragePlot(
  object = CHLA10,
  region = "TGFBI",
  # features = "TGFBI",
  expression.assay = "RNA",
  idents = idents.plot,
  extend.upstream = 500,
  extend.downstream = 10000
  )

peak_plot1 <- PeakPlot(
  object = CHLA10,
  region = goi_to_roi("TGFBI"),
  peaks = fli1_pks, sep = c("-","-"), 
  extend.upstream = 500,
  extend.downstream = 10000
)

peak_plot2 <- PeakPlot(
  object = CHLA10,
  region = goi_to_roi("TGFBI"),
  peaks = da_pks, sep = c("-","-"), 
  extend.upstream = 500,
  extend.downstream = 10000
)


```


```{r}
CombineTracks(plotlist = list(covplot, peak_plot1, peak_plot2))
```




