---
title: "a673_zeb_check"
output: html_document
---

```{r, message=F, echo=F}
library(Seurat)
library(scCustomize)
library(ggplot2)
library(dplyr)
library(viridis)
# library(SeuratData)
library(magrittr)
library(patchwork)
library(Signac)
# library(qs)
library(clusterProfiler)
library(msigdbr)
library(enrichplot)
library(RColorBrewer)
library(BSgenome.Hsapiens.UCSC.hg38)
library(GenomeInfoDb)
library(plyranges)
library(GenomicRanges)
library(EnsDb.Hsapiens.v86)
library(SeuratDisk)
library(SeuratObject)
library(reshape2)
library(eulerr)
library(org.Hs.eg.db)

```


```{r}#
# read in data
A673 <- readRDS(file = "../working_data/a673_tgfb_seurat.rds")
Idents(A673) <- "condition"
table(Idents(A673))
```

```{r}
# read in gene sets
riggi_up <- read.delim(file="../gene_lists/RIGGI_UP.txt", header=FALSE)
riggi_dn <- read.delim(file="../gene_lists/RIGGI_DN.txt", header=FALSE)

cc.genes <- Seurat::cc.genes.updated.2019

```


```{r}#
# first look at data
treatpal <- c("aquamarine2", "orange", "purple1")
DimPlot_scCustom(A673, colors_use = treatpal, label = F, reduction = "pca")
DimPlot_scCustom(A673, colors_use = treatpal, label = F, reduction = "umap")
DimPlot_scCustom(A673, colors_use = treatpal, label = F, reduction = "lsi")
DimPlot_scCustom(A673, colors_use = treatpal, label = F, reduction = "umap.atac")
DimPlot_scCustom(A673, colors_use = treatpal, label = F, reduction = "wnn.umap")
# cells kept their old clusters, let's try changing that
```

```{r}#
# take cell cycle out of the mix

# Score cell cycle for RNA assay
A673 <- CellCycleScoring(
  object = A673,
  s.features = cc.genes$s.genes,
  g2m.features = cc.genes$g2m.genes,
  set.ident = TRUE  # Optional: sets cell identity to phase
)

A673 <- ScaleData(A673, vars.to.regress = c("S.Score", "G2M.Score"), features = rownames(A673))

```


```{r}#
# re-sort RNA data
# A673 <- NormalizeData(object = A673)
A673 <- FindVariableFeatures(object = A673)
# A673 <- ScaleData(object = A673)
A673 <- RunPCA(object = A673)
ElbowPlot(A673, ndims = 50) # check this out tomorrow to check dims
A673 <- FindNeighbors(object = A673, dims = 1:30)
A673 <- FindClusters(object = A673)
A673 <- RunUMAP(object = A673, dims = 1:30)
```

```{r}#
# re-sort ATAC data
DefaultAssay(A673) <- "ATAC"
A673 <- RunTFIDF(A673)
A673 <- FindTopFeatures(A673, min.cutoff = 'q0')
A673 <- RunSVD(A673)
ElbowPlot(A673, reduction = "lsi", ndims = 50)
A673 <- RunUMAP(A673, reduction = 'lsi', dims = 2:20, reduction.name = "umap.atac", reduction.key = "atacUMAP_")
```

```{r}#
# cluster based on both RNA and ATAC data
A673 <- FindMultiModalNeighbors(A673, reduction.list = list("pca", "lsi"), dims.list = list(1:30, 2:20))
A673 <- RunUMAP(A673, nn.name = "weighted.nn", reduction.name = "wnn.umap", reduction.key = "wnnUMAP_")
A673 <- FindClusters(A673, graph.name = "wsnn", algorithm = 3, verbose = FALSE)
saveRDS(A673, "../working_data/A673_reclustered.rds")
```

## Re-processed data, start here

```{r}
A673 <- readRDS("../working_data/A673_reclustered.rds")
Idents(A673) <- "condition"
DefaultAssay(A673) <- "RNA"
# B673 <- readRDS("../working_data/A673_cc_reclustered.rds")
# DefaultAssay(B673) <- "RNA"

```


```{r}#
DimPlot(A673, reduction = "wnn.umap", group.by = "Phase", pt.size = 0.5) + ggtitle("Cell Cycle Phase on WNN UMAP") # need to add the phase idents, code above
DimPlot(B673, reduction = "wnn.umap", group.by = "Phase", pt.size = 0.5) + ggtitle("Cell Cycle Phase on WNN UMAP")
var_ftrs <- VariableFeatures(A673)
var_ftrs_cc <- VariableFeatures(B673)
checker <- intersect(var_ftrs, var_ftrs_cc)
```

```{r}
treatpal <- c("aquamarine2", "orange", "purple1")

DimPlot_scCustom(A673, colors_use = treatpal, label = F, reduction = "pca")
DimPlot_scCustom(A673, colors_use = treatpal, label = F, reduction = "umap")
DimPlot_scCustom(A673, colors_use = treatpal, label = F, reduction = "lsi")
DimPlot_scCustom(A673, colors_use = treatpal, label = F, reduction = "umap.atac")
DimPlot_scCustom(A673, colors_use = treatpal, label = F, reduction = "wnn.umap")
Cluster_Highlight_Plot(A673, c("vehicle", "TGFB1", "TGFB2"), highlight_color = c("red","blue1","blue4"), reduction = "wnn.umap")

```

```{r}
# adding metadata and idents based on riggi gene lists
A673 <- AddModuleScore(A673, features=riggi_up, name="riggi_up")
A673 <- AddModuleScore(A673, features=riggi_dn, name="riggi_dn")

# breaking riggi distributions
rup_cutoff <- quantile(A673$riggi_up1, probs=0.5)
rup_cutoff <- unname(rup_cutoff)

rdn_cutoff <- quantile(A673$riggi_dn1, probs=0.5)
rdn_cutoff <- unname(rdn_cutoff)
```

```{r}
# adding riggi idents

riggi_up_hi <- subset(x = A673, subset = riggi_up1 > rup_cutoff, slot='data')
riggi_up_lo <- subset(x = A673, subset = riggi_up1 > rup_cutoff, slot='data', invert=TRUE)
A673 <- SetIdent(A673, cells=Cells(riggi_up_hi), value='riggi_up_hi')
A673 <- SetIdent(A673, cells=Cells(riggi_up_lo), value='riggi_up_lo')
A673$riggi_up_identity <- Idents(A673)
remove(riggi_up_hi)
remove(riggi_up_lo)

riggi_dn_hi <- subset(x = A673, subset = riggi_dn1 > rdn_cutoff, slot='data')
riggi_dn_lo <- subset(x = A673, subset = riggi_dn1 > rdn_cutoff, slot='data', invert=TRUE)
A673 <- SetIdent(A673, cells=Cells(riggi_dn_hi), value='riggi_dn_hi')
A673 <- SetIdent(A673, cells=Cells(riggi_dn_lo), value='riggi_dn_lo')
A673$riggi_dn_identity <- Idents(A673)
remove(riggi_dn_hi)
remove(riggi_dn_lo)
```

```{r}
# further sorting cells by riggi levels
hybrid<- subset(x = A673, subset = riggi_up_identity  == "riggi_up_hi" & riggi_dn_identity == "riggi_dn_hi", slot='data')
true_low<- subset(x = A673, subset = riggi_up_identity  == "riggi_up_lo" & riggi_dn_identity == "riggi_dn_hi", slot='data')
true_hi<- subset(x = A673, subset = riggi_up_identity  == "riggi_up_hi" & riggi_dn_identity == "riggi_dn_lo", slot='data')
low_hybrid<- subset(x = A673, subset = riggi_up_identity  == "riggi_up_lo" & riggi_dn_identity == "riggi_dn_lo", slot='data')

A673 <- SetIdent(A673, cells=Cells(hybrid), value='hybrid')
A673 <- SetIdent(A673, cells=Cells(true_low), value='true_low')
A673 <- SetIdent(A673, cells=Cells(true_hi), value='true_high')
A673 <- SetIdent(A673, cells=Cells(low_hybrid), value='low_hybrid')
Idents(A673) <- factor(Idents(A673), levels = c("low_hybrid", "true_low", "true_high", "hybrid"))

A673$ef_state_identity <- Idents(A673)

remove(hybrid)
remove(true_low)
remove(true_hi)
remove(low_hybrid)

```



```{r}
pal <- c("#FEC98DFF" ,"#FD9567FF", "#F1605DFF", "#CD4071FF" ,"#9F2F7FFF" ,"#721F81FF", "#451077FF")
FeaturePlot_scCustom(A673, features="ZEB1", colors_use=pal, reduction = "wnn.umap")
FeaturePlot_scCustom(A673, features="ZEB2", colors_use=pal, reduction = "wnn.umap")
FeaturePlot_scCustom(A673, features="FOSL2", colors_use=pal, reduction = "wnn.umap")
FeaturePlot_scCustom(A673, features="TGFBI", colors_use=pal, reduction = "wnn.umap")
FeaturePlot_scCustom(A673, features="MKI67", colors_use=pal, reduction = "wnn.umap")

```

```{r}
Idents(A673) <- "condition"
DotPlot_scCustom(A673, features=c("ZEB1", "ZEB2", "FOSL2", "TGFBI"), scale.min=0, dot.scale=20, scale=TRUE, colors_use=inferno(100), group.by="condition") + ggtitle("A673 single cell treatments")
```

```{r}
# plot riggi up and down module scores by condition - sccustomize doesn't work on module scores

```

# subsetting cells by zeb2

```{r}
zeb2_expr <- GetAssayData(A673, slot = "data")["ZEB2", ]

# quantile(FetchData(A673, vars = "ZEB2")[, 1], probs = seq(0,1,0.1))
# zeb2_bot_30 <- quantile(zeb2_expr, 0.30)
# zeb2_top_30 <- quantile(zeb2_expr, 0.70)
# 
# zeb2_bot_30 <- unname(zeb2_bot_30)
# zeb2_top_30 <- unname(zeb2_top_30)
```

```{r}
ZEB2_positive <- subset(A673, cells = names(zeb2_expr[zeb2_expr > 0]))
ZEB2_negative <- subset(A673, cells = names(zeb2_expr[zeb2_expr <= 0]))


# this code also sets the active identity of the cells to these categories
A673 <- SetIdent(A673, cells=Cells(ZEB2_positive), value='ZEB2+')
A673 <- SetIdent(A673, cells=Cells(ZEB2_negative), value='ZEB2-')

# make sure they are in an appropriate order before saving them to object
A673$zeb2_expr <- Idents(A673)

remove(ZEB2_positive)
remove(ZEB2_negative)
```

```{r}
# dot plots
DotPlot(
  object = A673,
  features = c("riggi_up1", "riggi_dn1"),
  group.by = "condition"
) +
  scale_color_viridis(option = "inferno") + scale_size(range = c(2, 11)) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  ggtitle("A673 riggi genesets by treatment")

DotPlot(
  object = A673,
  features = c("riggi_up1", "riggi_dn1"),
  group.by = "zeb2_expr"
) +
  scale_color_viridis(option = "inferno") + scale_size(range = c(2, 10)) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  ggtitle("A673 riggi genesets by zeb2")

DotPlot(
  object = A673,
  features = c("FOSL2", "JUND", "ZEB1", "TGFBI"),
  group.by = "ef_state_identity"
) +
  scale_color_viridis(option = "inferno") + scale_size(range = c(2, 10)) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  ggtitle("A673 GOI by riggi geneset expression levels")

DotPlot(
  object = A673,
  features = c("FOSL2", "JUND", "ZEB1", "TGFBI"),
  group.by = "zeb2_expr"
) +
  scale_color_viridis(option = "inferno") + scale_size(range = c(2, 10)) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  ggtitle("A673 GOI by ZEB2 expression groups")

```

```{r}
p1 <- DimPlot(A673, reduction = "umap", group.by = "ef_state_identity", label = TRUE, label.size = 2.5, repel = TRUE) + ggtitle("RNA") + NoLegend()
p2 <- DimPlot(A673, reduction = "umap.atac", group.by = "ef_state_identity", label = TRUE, label.size = 2.5, repel = TRUE) + ggtitle("ATAC") + NoLegend()
p3 <- DimPlot(A673, reduction = "wnn.umap", group.by = "ef_state_identity", label = TRUE, label.size = 2.5, repel = TRUE) + ggtitle("WNN")
p1 + p2 + p3 & theme(plot.title = element_text(hjust = 0.5))
```

```{r}
# fragment file still not indexed
# Idents(A673) <- 
# CoveragePlot(A673, region = 'CD8A', assay = 'ATAC', expression.assay = 'RNA', peaks = FALSE)
# colnames(A673[["ATAC"]])
```

```{r}
# reset annotations on Seurat object, they are wonky
annotations <- GetGRangesFromEnsDb(ensdb = EnsDb.Hsapiens.v86)
seqlevelsStyle(annotations) <- "UCSC"
genome(annotations) <- "hg38"
Annotation(A673[["ATAC"]]) <- annotations
```


```{r}
# Identify DA peaks
DefaultAssay(A673) <- 'ATAC'
Idents(A673) <- "condition"

# wilcox is the default option for test.use
tgfb1_da_peaks <- FindMarkers(
  object = A673,
  ident.1 = "TGFB1",
  ident.2 = "vehicle",
  test.use = 'wilcox',
  min.pct = 0.1
)
```

```{r}
# record significantly DA peaks
open_tgfb1 <- rownames(tgfb1_da_peaks[tgfb1_da_peaks$avg_log2FC > 0 & tgfb1_da_peaks$p_val_adj < 0.05, ])
open_vehicle <- rownames(tgfb1_da_peaks[tgfb1_da_peaks$avg_log2FC < 0 & tgfb1_da_peaks$p_val_adj < 0.05, ])
```

```{r}
# find genes associated with DA peaks
closest_genes_tgfb1 <- ClosestFeature(A673, regions = open_tgfb1)
closest_genes_vehicle <- ClosestFeature(A673, regions = open_vehicle) # somehow no DA areas found?

A673_tgfb1_accessible_genes <- closest_genes_tgfb1$gene_name
A673_tgfb1_inaccessible_genes <- closest_genes_vehicle$gene_name
```

```{r}
# import rna assay marker genes for the same contrast
markergenes_A673_tgfb1_veh <- read.csv("../../VTP_SC_seq/data/A673_tgfb1_vs_veh.csv")
tgfb1_upreg <- markergenes_A673_tgfb1_veh[markergenes_A673_tgfb1_veh$avg_log2FC > 0 & markergenes_A673_tgfb1_veh$p_val_adj < 0.05, "X"]

tgfb1_dnreg <- markergenes_A673_tgfb1_veh[markergenes_A673_tgfb1_veh$avg_log2FC < 0 & markergenes_A673_tgfb1_veh$p_val_adj < 0.05, "X"]


```

```{r}
tgfb1_vs_veh_multi_upregs <- list(ATAC = unique(A673_tgfb1_accessible_genes), RNA = unique(tgfb1_upreg))

# Convert to eulerr-compatible format
fit <- euler(tgfb1_vs_veh_multi_upregs)
plot(fit, fills = c("red", "yellow"), quantities = TRUE, main = "A673 tgfb1_vs_veh multi upreg overlaps")

tgfb1_vs_veh_multi_dnregs <- list(ATAC = unique(A673_tgfb1_inaccessible_genes), RNA = unique(tgfb1_dnreg))

# Convert to eulerr-compatible format
fit <- euler(tgfb1_vs_veh_multi_dnregs)
plot(fit, fills = c("red", "yellow"), quantities = TRUE, main = "A673 tgfb1_vs_veh multi downreg overlaps")
```

```{r}
# bring in kate's bulk data too
bulk_A673_tgfb1_vs_veh <- read.csv("../../VTP_SC_seq/bulk_data/210208-TGFB1vsVehicle_A673.csv")

bulk_A673_tgfb1_vs_veh_filtered <- bulk_A673_tgfb1_vs_veh %>% dplyr::filter(!is.na(padj) & padj < 0.05)
bulk_A673_tgfb1_vs_veh_upreg <- bulk_A673_tgfb1_vs_veh_filtered$SYMBOL[bulk_A673_tgfb1_vs_veh_filtered$log2FoldChange > 0]
bulk_A673_tgfb1_vs_veh_downreg <- bulk_A673_tgfb1_vs_veh_filtered$SYMBOL[bulk_A673_tgfb1_vs_veh_filtered$log2FoldChange < 0]
```

```{r}
tgfb1_vs_veh_multi_more_upregs <- list(ATAC = unique(A673_tgfb1_accessible_genes), RNA = unique(tgfb1_upreg), BULK = unique(bulk_A673_tgfb1_vs_veh_upreg))

# Convert to eulerr-compatible format
fit <- euler(tgfb1_vs_veh_multi_more_upregs)
plot(fit, fills = c("red", "yellow", "blue"), quantities = TRUE, main = "A673 tgfb1_vs_veh multi upreg overlaps")

tgfb1_vs_veh_multi_more_dnregs <- list(ATAC = unique(A673_tgfb1_inaccessible_genes), RNA = unique(tgfb1_dnreg), BULK = unique(bulk_A673_tgfb1_vs_veh_downreg))

# Convert to eulerr-compatible format
fit <- euler(tgfb1_vs_veh_multi_more_dnregs)
plot(fit, fills = c("red", "yellow", "blue"), quantities = TRUE, main = "A673 tgfb1_vs_veh multi downreg overlaps")

```



```{r}
tgfb1_up_overlaps <- intersect(A673_tgfb1_accessible_genes, tgfb1_upreg)

tgfb1_up_overlaps
```

```{r}
tgfb1_dn_overlaps <- intersect(A673_tgfb1_inaccessible_genes, tgfb1_dnreg)

tgfb1_dn_overlaps

```

```{r}
# run ORA for overlapping genes
t2g_hallmarks <- msigdbr(species = "Homo sapiens", collection ="H") %>%
  dplyr::select(gs_name, gene_symbol)
t2g_reactome <- msigdbr(species = "Homo sapiens", collection ="C2", subcollection = "CP:REACTOME") %>%
  dplyr::select(gs_name, gene_symbol)

tgfb1_dn_GOALL <- enrichGO(tgfb1_dn_overlaps, OrgDb = org.Hs.eg.db, keyType = "SYMBOL", ont = "ALL", minGSSize = 10, pvalueCutoff = 1)
tgfb1_dn_hallmarks <- enricher(tgfb1_dn_overlaps, TERM2GENE = t2g_hallmarks)
tgfb1_dn_react <- enricher(tgfb1_dn_overlaps, TERM2GENE = t2g_reactome)

```

```{r}
# overlap gene set is too small to have significant results
# dotplot(tgfb1_dn_GOALL, title = "A673 TGFB1-veh upregulated overlaps GOALL")
# dotplot(tgfb1_dn_hallmarks, title = "A673 TGFB1-veh upregulated overlaps hallmarks")
# dotplot(tgfb1_dn_react, title = "A673 TGFB1-veh upregulated overlaps reactome")

```

