---
title: "chla10_zeb_check"
output: html_document
---

```{r, message=F, echo=F}
library(Seurat)
library(scCustomize)
library(ggplot2)
library(dplyr)
library(viridis)
# library(SeuratData)
library(magrittr)
library(patchwork)
library(Signac)
# library(qs)
library(clusterProfiler)
library(msigdbr)
library(enrichplot)
library(RColorBrewer)
library(BSgenome.Hsapiens.UCSC.hg38)
library(GenomeInfoDb)
library(plyranges)
library(GenomicRanges)
library(EnsDb.Hsapiens.v86)
library(SeuratDisk)
library(SeuratObject)
library(reshape2)
library(eulerr)
library(org.Hs.eg.db)

```


```{r}#
# read in data
CHLA10 <- readRDS(file = "../working_data/chla10_tgfb_seurat.rds")
Idents(CHLA10) <- "condition"
table(Idents(CHLA10))
```

```{r}
# read in gene sets
riggi_up <- read.delim(file="../gene_lists/RIGGI_UP.txt", header=FALSE)
riggi_dn <- read.delim(file="../gene_lists/RIGGI_DN.txt", header=FALSE)

cc.genes <- Seurat::cc.genes.updated.2019

```


```{r}#
# first look at data
treatpal <- c("aquamarine2", "orange", "purple1")
DimPlot_scCustom(CHLA10, colors_use = treatpal, label = F, reduction = "pca")
DimPlot_scCustom(CHLA10, colors_use = treatpal, label = F, reduction = "umap")
DimPlot_scCustom(CHLA10, colors_use = treatpal, label = F, reduction = "lsi")
DimPlot_scCustom(CHLA10, colors_use = treatpal, label = F, reduction = "umap.atac")
DimPlot_scCustom(CHLA10, colors_use = treatpal, label = F, reduction = "wnn.umap")
# cells kept their old clusters, let's try changing that
```

```{r}
# re-sort RNA data
# CHLA10 <- NormalizeData(object = CHLA10)
CHLA10 <- FindVariableFeatures(object = CHLA10)
CHLA10 <- ScaleData(object = CHLA10)
CHLA10 <- RunPCA(object = CHLA10)
ElbowPlot(CHLA10, ndims = 50) # check this out tomorrow to check dims
CHLA10 <- FindNeighbors(object = CHLA10, dims = 1:30)
CHLA10 <- FindClusters(object = CHLA10)
CHLA10 <- RunUMAP(object = CHLA10, dims = 1:30)
```

```{r}
# re-sort ATAC data
DefaultAssay(CHLA10) <- "ATAC"
CHLA10 <- RunTFIDF(CHLA10)
CHLA10 <- FindTopFeatures(CHLA10, min.cutoff = 'q0')
CHLA10 <- RunSVD(CHLA10)
ElbowPlot(CHLA10, reduction = "lsi", ndims = 50)
CHLA10 <- RunUMAP(CHLA10, reduction = 'lsi', dims = 2:20, reduction.name = "umap.atac", reduction.key = "atacUMAP_")
```

```{r}#
# cluster based on both RNA and ATAC data
CHLA10 <- FindMultiModalNeighbors(CHLA10, reduction.list = list("pca", "lsi"), dims.list = list(1:30, 2:20))
CHLA10 <- RunUMAP(CHLA10, nn.name = "weighted.nn", reduction.name = "wnn.umap", reduction.key = "wnnUMAP_")
CHLA10 <- FindClusters(CHLA10, graph.name = "wsnn", algorithm = 3, verbose = FALSE)
# saveRDS(CHLA10, "../working_data/CHLA10_reclustered.rds")
```

## Re-processed data, start here

```{r}
CHLA10 <- readRDS("../working_data/CHLA10_reclustered.rds")
DefaultAssay(CHLA10) <- "RNA"

# Score cell cycle for RNA assay
CHLA10 <- CellCycleScoring(
  object = CHLA10,
  s.features = cc.genes$s.genes,
  g2m.features = cc.genes$g2m.genes,
  set.ident = TRUE  # Optional: sets cell identity to phase
)
```


```{r}
Idents(CHLA10) <- "condition"
treatpal <- c("aquamarine2", "orange", "purple1")

DimPlot_scCustom(CHLA10, colors_use = treatpal, label = F, reduction = "pca")
DimPlot_scCustom(CHLA10, colors_use = treatpal, label = F, reduction = "umap")
DimPlot_scCustom(CHLA10, colors_use = treatpal, label = F, reduction = "lsi")
DimPlot_scCustom(CHLA10, colors_use = treatpal, label = F, reduction = "umap.atac")
DimPlot_scCustom(CHLA10, colors_use = treatpal, label = F, reduction = "wnn.umap")
Cluster_Highlight_Plot(CHLA10, c("vehicle", "TGFB1", "TGFB2"), highlight_color = c("red","blue1","blue4"), reduction = "wnn.umap")

```

```{r}
# adding metadata and idents based on riggi gene lists

CHLA10 <- AddModuleScore(CHLA10, features=riggi_up, name="riggi_up")
CHLA10 <- AddModuleScore(CHLA10, features=riggi_dn, name="riggi_dn")

# breaking riggi distributions
rup_cutoff <- quantile(CHLA10$riggi_up1, probs=0.5)
rup_cutoff <- unname(rup_cutoff)

rdn_cutoff <- quantile(CHLA10$riggi_dn1, probs=0.5)
rdn_cutoff <- unname(rdn_cutoff)
```

```{r}
# adding riggi idents

riggi_up_hi <- subset(x = CHLA10, subset = riggi_up1 > rup_cutoff, slot='data')
riggi_up_lo <- subset(x = CHLA10, subset = riggi_up1 > rup_cutoff, slot='data', invert=TRUE)
CHLA10 <- SetIdent(CHLA10, cells=Cells(riggi_up_hi), value='riggi_up_hi')
CHLA10 <- SetIdent(CHLA10, cells=Cells(riggi_up_lo), value='riggi_up_lo')
CHLA10$riggi_up_identity <- Idents(CHLA10)
remove(riggi_up_hi)
remove(riggi_up_lo)

riggi_dn_hi <- subset(x = CHLA10, subset = riggi_dn1 > rdn_cutoff, slot='data')
riggi_dn_lo <- subset(x = CHLA10, subset = riggi_dn1 > rdn_cutoff, slot='data', invert=TRUE)
CHLA10 <- SetIdent(CHLA10, cells=Cells(riggi_dn_hi), value='riggi_dn_hi')
CHLA10 <- SetIdent(CHLA10, cells=Cells(riggi_dn_lo), value='riggi_dn_lo')
CHLA10$riggi_dn_identity <- Idents(CHLA10)
remove(riggi_dn_hi)
remove(riggi_dn_lo)
```

```{r}
# further sorting cells by riggi levels
hybrid<- subset(x = CHLA10, subset = riggi_up_identity  == "riggi_up_hi" & riggi_dn_identity == "riggi_dn_hi", slot='data')
true_low<- subset(x = CHLA10, subset = riggi_up_identity  == "riggi_up_lo" & riggi_dn_identity == "riggi_dn_hi", slot='data')
true_hi<- subset(x = CHLA10, subset = riggi_up_identity  == "riggi_up_hi" & riggi_dn_identity == "riggi_dn_lo", slot='data')
low_hybrid<- subset(x = CHLA10, subset = riggi_up_identity  == "riggi_up_lo" & riggi_dn_identity == "riggi_dn_lo", slot='data')

CHLA10 <- SetIdent(CHLA10, cells=Cells(hybrid), value='hybrid')
CHLA10 <- SetIdent(CHLA10, cells=Cells(true_low), value='true_low')
CHLA10 <- SetIdent(CHLA10, cells=Cells(true_hi), value='true_high')
CHLA10 <- SetIdent(CHLA10, cells=Cells(low_hybrid), value='low_hybrid')
Idents(CHLA10) <- factor(Idents(CHLA10), levels = c("low_hybrid", "true_low", "true_high", "hybrid"))

CHLA10$ef_state_identity <- Idents(CHLA10)

remove(hybrid)
remove(true_low)
remove(true_hi)
remove(low_hybrid)

```



```{r}
pal <- c("#FEC98DFF" ,"#FD9567FF", "#F1605DFF", "#CD4071FF" ,"#9F2F7FFF" ,"#721F81FF", "#451077FF")
FeaturePlot_scCustom(CHLA10, features="ZEB1", colors_use=pal, reduction = "wnn.umap")
FeaturePlot_scCustom(CHLA10, features="ZEB2", colors_use=pal, reduction = "wnn.umap")
FeaturePlot_scCustom(CHLA10, features="FOSL2", colors_use=pal, reduction = "wnn.umap")
FeaturePlot_scCustom(CHLA10, features="TGFBI", colors_use=pal, reduction = "wnn.umap")
FeaturePlot_scCustom(CHLA10, features="MKI67", colors_use=pal, reduction = "wnn.umap")

```

```{r}
Idents(CHLA10) <- "condition"
DotPlot_scCustom(CHLA10, features=c("ZEB1", "ZEB2", "FOSL2", "TGFBI"), scale.min=0, dot.scale=20, scale=TRUE, colors_use=inferno(100), group.by="condition") + ggtitle("CHLA10 single cell treatments")
```

```{r}
# cell cycle dotplots
Idents(CHLA10) <- "Phase"
DotPlot_scCustom(CHLA10, features=c("ZEB1", "ZEB2"), scale.min=0, dot.scale=20, scale=TRUE, colors_use=inferno(100), group.by="Phase") + ggtitle("CHLA10 cell phases")
DotPlot_scCustom(CHLA10, features=c("AXL", "CDKN1B", "CDKN1C"), scale.min=0, dot.scale=20, scale=TRUE, colors_use=inferno(100), group.by="Phase") + ggtitle("CHLA10 cell phases")
```

# subsetting cells by zeb2

```{r}
zeb2_expr <- GetAssayData(CHLA10, slot = "data")["ZEB2", ]

# quantile(FetchData(CHLA10, vars = "ZEB2")[, 1], probs = seq(0,1,0.1))
zeb2_bot_30 <- quantile(zeb2_expr, 0.30)
zeb2_top_30 <- quantile(zeb2_expr, 0.70)

zeb2_bot_30 <- unname(zeb2_bot_30)
zeb2_top_30 <- unname(zeb2_top_30)
```

```{r}
hi_ZEB2 <- subset(CHLA10, cells = names(zeb2_expr[zeb2_expr > zeb2_top_30]))
low_ZEB2 <- subset(CHLA10, cells = names(zeb2_expr[zeb2_expr < zeb2_bot_30]))

mid_zeb2_cells <- names(zeb2_expr[zeb2_expr >= zeb2_bot_30 & zeb2_expr <= zeb2_top_30])
mid_ZEB2 <- subset(CHLA10, cells = mid_zeb2_cells)

# this code also sets the active identity of the cells to these categories
CHLA10 <- SetIdent(CHLA10, cells=Cells(hi_ZEB2), value='Hi_ZEB2')
CHLA10 <- SetIdent(CHLA10, cells=Cells(mid_ZEB2), value='Mid_ZEB2')
CHLA10 <- SetIdent(CHLA10, cells=Cells(low_ZEB2), value='Low_ZEB2')

# make sure they are in an appropriate order before saving them to object
Idents(CHLA10) <- factor(Idents(CHLA10), levels = c("Low_ZEB2", "Mid_ZEB2", "Hi_ZEB2"))
CHLA10$zeb2_lvl <- Idents(CHLA10)

remove(hi_ZEB2)
remove(mid_ZEB2)
remove(low_ZEB2)
```

```{r}
# dot plots
DotPlot(
  object = CHLA10,
  features = c("riggi_up1", "riggi_dn1"),
  group.by = "condition"
) +
  scale_color_viridis(option = "inferno") + scale_size(range = c(2, 11)) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  ggtitle("CHLA10 riggi genesets by treatment")

DotPlot(
  object = CHLA10,
  features = c("riggi_up1", "riggi_dn1"),
  group.by = "zeb2_lvl"
) +
  scale_color_viridis(option = "inferno") + scale_size(range = c(2, 10)) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  ggtitle("CHLA10 riggi genesets by zeb2")

DotPlot(
  object = CHLA10,
  features = c("FOSL2", "JUND", "ZEB1", "TGFBI"),
  group.by = "ef_state_identity"
) +
  scale_color_viridis(option = "inferno") + scale_size(range = c(2, 10)) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  ggtitle("CHLA10 GOI by riggi geneset expression levels")

DotPlot(
  object = CHLA10,
  features = c("FOSL2", "JUND", "ZEB1", "TGFBI"),
  group.by = "zeb2_lvl"
) +
  scale_color_viridis(option = "inferno") + scale_size(range = c(2, 10)) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  ggtitle("CHLA10 GOI by ZEB2 expression groups")

```

```{r}
p1 <- DimPlot(CHLA10, reduction = "umap", group.by = "ef_state_identity", label = TRUE, label.size = 2.5, repel = TRUE) + ggtitle("RNA") + NoLegend()
p2 <- DimPlot(CHLA10, reduction = "umap.atac", group.by = "ef_state_identity", label = TRUE, label.size = 2.5, repel = TRUE) + ggtitle("ATAC") + NoLegend()
p3 <- DimPlot(CHLA10, reduction = "wnn.umap", group.by = "ef_state_identity", label = TRUE, label.size = 2.5, repel = TRUE) + ggtitle("WNN")
p1 + p2 + p3 & theme(plot.title = element_text(hjust = 0.5))
```

```{r}
# fragment file still not indexed
# Idents(CHLA10) <- 
# CoveragePlot(CHLA10, region = 'CD8A', assay = 'ATAC', expression.assay = 'RNA', peaks = FALSE)
# colnames(CHLA10[["ATAC"]])
```

```{r}
# reset annotations on Seurat object, they are wonky -- check and see if Olivia's code can replace this
annotations <- GetGRangesFromEnsDb(ensdb = EnsDb.Hsapiens.v86)
seqlevelsStyle(annotations) <- "UCSC"
genome(annotations) <- "hg38"
Annotation(CHLA10[["ATAC"]]) <- annotations
```


```{r}
# Identify DA peaks
DefaultAssay(CHLA10) <- 'ATAC'
Idents(CHLA10) <- 'condition'


# wilcox is the default option for test.use
tgfb1_da_peaks <- FindMarkers(
  object = CHLA10,
  ident.1 = "TGFB1",
  ident.2 = "vehicle",
  test.use = 'wilcox',
  min.pct = 0.1
)
```

```{r}
# record significantly DA peaks
open_tgfb1 <- rownames(tgfb1_da_peaks[tgfb1_da_peaks$avg_log2FC > 0 & tgfb1_da_peaks$p_val_adj < 0.05, ])
open_vehicle <- rownames(tgfb1_da_peaks[tgfb1_da_peaks$avg_log2FC < 0 & tgfb1_da_peaks$p_val_adj < 0.05, ])
```

```{r}
# find genes associated with DA peaks
closest_genes_tgfb1 <- ClosestFeature(CHLA10, regions = open_tgfb1)
# closest_genes_vehicle <- ClosestFeature(CHLA10, regions = open_vehicle) # somehow no DA areas found?

tgfb1_accessible_genes <- closest_genes_tgfb1$gene_name
```

```{r}
# import rna assay marker genes for the same contrast
markergenes_CHLA_tgfb1_veh <- read.csv("../../VTP_SC_seq/data/CHLA10_tgfb1_vs_veh.csv")
tgfb1_upreg <- markergenes_CHLA_tgfb1_veh[markergenes_CHLA_tgfb1_veh$avg_log2FC > 0 & markergenes_CHLA_tgfb1_veh$p_val_adj < 0.05, "X"]
```

```{r}
tgfb1_vs_veh_multi_upregs <- list(ATAC = unique(tgfb1_accessible_genes), RNA = unique(tgfb1_upreg))

# Convert to eulerr-compatible format
fit <- euler(tgfb1_vs_veh_multi_upregs)
plot(fit, fills = c("red", "yellow"), quantities = TRUE, main = "CHLA10 tgfb1_vs_veh multiomic overlaps - upreg")


```


```{r}
tgfb1_up_overlaps <- intersect(tgfb1_accessible_genes, tgfb1_upreg)

tgfb1_up_overlaps
```

```{r}
# run ORA for overlapping genes
t2g_hallmarks <- msigdbr(species = "Homo sapiens", collection ="H") %>%
  dplyr::select(gs_name, gene_symbol)
t2g_reactome <- msigdbr(species = "Homo sapiens", collection ="C2", subcollection = "CP:REACTOME") %>%
  dplyr::select(gs_name, gene_symbol)

tgfb1_up_GOALL <- enrichGO(tgfb1_up_overlaps, OrgDb = org.Hs.eg.db, keyType = "SYMBOL", ont = "ALL", minGSSize = 10, pvalueCutoff = 1)
tgfb1_up_hallmarks <- enricher(tgfb1_up_overlaps, TERM2GENE = t2g_hallmarks)
tgfb1_up_react <- enricher(tgfb1_up_overlaps, TERM2GENE = t2g_reactome)

```

```{r}
dotplot(tgfb1_up_GOALL, title = "CHLA10 TGFB1-veh upregulated overlaps GOALL")
dotplot(tgfb1_up_hallmarks, title = "CHLA10 TGFB1-veh upregulated overlaps hallmarks")
dotplot(tgfb1_up_react, title = "CHLA10 TGFB1-veh upregulated overlaps reactome")

```


```{r}
# link peaks
CHLA10 <- LinkPeaks(object = CHLA10, peak.assay = "ATAC", expression.assay = "RNA") # takes time!

CHLA10_links <-  Links(CHLA10)

# write.csv(CHLA10_links, "../results/CHLA10_link_ranges.csv")
```


```{r}
# motif enrichment
upreg_pks <- tgfb1_da_peaks[tgfb1_da_peaks$avg_log2FC > 0 & tgfb1_da_peaks$p_val_adj < 0.05, ]

# ensure peaks are compatible with your seurat object
valid_peaks <- intersect(as.character(rownames(upreg_pks)), rownames(CHLA10[["ATAC"]]))

DefaultAssay(CHLA10) <- "ATAC"

enriched_motifs <- FindMotifs(CHLA10, features = valid_peaks) 
# finds overrepresented motifs in select peaks, such as upregulated (DA) peaks

MotifPlot(CHLA10, head(rownames(enriched_motifs)))
```

```{r}
# DA motifs
saveRDS(CHLA10, "../working_data/CHLA10_linked.rds")
# CHLA10 <- RunChromVAR(CHLA10, genome = BSgenome.Hsapiens.UCSC.hg38) # takes long time!

```




